### PR TITLE
Feat/block upgrade if vault in sealed

### PR DESCRIPTION
**Pull request by jneo8**
Add pre-check step as first step global before all the steps to make sure vault is not in sealed.

follow: #507 
relate: #467

### CODE REVIEW

File: cou/steps/vault.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    :type model: Model
+    :raises VaultSealed: if application in sealed status
+    """
+    try:
+        app_names = await model.get_application_names(charm_name="vault")
+        for app_name in app_names:
+            app = await model.get_application_status(app_name=app_name)
+            if app.status.info == "Unit is sealed" and app.status.status == "blocked":
+                raise VaultSealed(
+                    "Vault is in sealed, please follow the steps on "
```
Comment (chanchiwai-ray) [inline cou/steps/vault.py:36]: (non-blocking) nit: `Vault is sealed ...`

File: tests/functional/tests/smoke.py [hunk 2]
```diff
@@ -134,6 +134,7 @@ def generate_expected_plan(self, backup: bool = True) -> str:
         backup_plan = "\tBack up MySQL databases\n" if backup else ""
         return (
             "Upgrade cloud from 'ussuri' to 'victoria'\n"
+            "\tCheck application vault is not sealed\n"
```
Comment (chanchiwai-ray) [inline tests/functional/tests/smoke.py:137]: (non-blocking) nit: `Verify application vault is not sealed\n` just to be consistent

File: cou/steps/plan.py [hunk 3]
```diff
@@ -369,6 +370,11 @@ def _get_pre_upgrade_steps(analysis_result: Analysis, args: CLIargs) -> list[Pre
     :rtype: list[PreUpgradeStep]
     """
     steps = [
+        PreUpgradeStep(
+            description="Check application vault is not sealed",
```
Comment (samuelallan72) [inline cou/steps/plan.py:374]: The wording could be improved a bit:

```suggestion
            description="Verify vault application is unsealed",
```

**Review by chanchiwai-ray — APPROVED** [review]
Since it's been tested on Danny's personal OpenStack deployment

File: cou/steps/vault.py [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+import logging
+
+from cou.exceptions import ApplicationNotFound, VaultSealed
+from cou.utils.juju_utils import Model
+
+logger = logging.getLogger(__name__)
+
+
+async def check_vault_status(model: Model) -> None:
+    """Make sure vault is not in sealed status.
```
Comment (samuelallan72) [inline cou/steps/vault.py:24]: Just some docs suggestions:

- maybe name it something a bit clearer, like `verify_vault_is_unsealed`, or `validate_vault_is_unsealed`:

```suggestion
async def verify_vault_is_unsealed(model: Model) -> None:
    """Verify vault is unsealed.
    
    Check vault status. If vault is sealed, raise VaultSealed.
```

**Review by samuelallan72 — APPROVED** [review]
The logic looks good to me.  I left some suggestions for improving the docs though (function naming, user-facing strings, etc.)

File: cou/steps/vault.py (hunk 1 above)
Comment (jneo8) [reply cou/steps/vault.py]: Updated.

File: tests/functional/tests/smoke.py (hunk 2 above)
Comment (jneo8) [reply tests/functional/tests/smoke.py]: Updated.

File: cou/steps/plan.py (hunk 3 above)
Comment (jneo8) [reply cou/steps/plan.py]: Updated.

**Review by samuelallan72 — APPROVED** [review]

File: cou/steps/vault.py (hunk 4 above)
Comment (jneo8) [reply cou/steps/vault.py]: Updated.

**Review by samuelallan72 — APPROVED** [review]

**Review by chanchiwai-ray — APPROVED** [review]

**Pull request merged**